diff --git a/CMakeLists.txt b/CMakeLists.txt
index 0f06fbb41..4f1afc855 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -477,15 +477,8 @@ if (NOT NOZLIB AND NOT ((NOT BUILD_SHARED_LIBS) AND WIN32))
     ENDIF (ZLIB_FOUND)
 endif()
 
-
-find_library(cadical
-    PATHS ${CMAKE_CURRENT_SOURCE_DIR}/../cadical/build/
-    NAMES cadical
-    REQUIRED)
-find_library(cadiback
-    PATHS ${CMAKE_CURRENT_SOURCE_DIR}/../cadiback/
-    NAMES cadiback
-    REQUIRED)
+find_library(cadical NAMES cadical REQUIRED)
+find_library(cadiback NAMES cadiback REQUIRED)
 
 include(CheckFloatPrecision)
 check_float_precision()
