diff --git a/configure b/configure
index f8f1457..29834b0 100755
--- a/configure
+++ b/configure
@@ -34,16 +34,14 @@ checkfile () {
   [ -f "$1" ] || die "could not find '$1'"
 }
 
-[ x"$CADICAL" = x ] && CADICAL="../cadical"
+checkdir "$CADICAL_DIR"
+cadicalversion="$(cat "$CADICAL_DIR"/VERSION)"
 
-checkdir "$CADICAL"
-cadicalversion="`cat $CADICAL/VERSION`"
+checkdir "$CADICAL_DIR/src"
+checkdir "$CADICAL_DIR/build"
 
-checkdir "$CADICAL/src"
-checkdir "$CADICAL/build"
-
-checkfile "$CADICAL/src/cadical.hpp"
-checkfile "$CADICAL/build/libcadical.a"
+checkfile "$CADICAL_DIR/src/cadical.hpp"
+checkfile "$CADICAL_DIR/build/libcadical.a"
 
 while [ $# -gt 0 ]
 do
@@ -56,9 +54,9 @@ do
   shift
 done
 
-msg "Using CADICAL='$CADICAL' version '$cadicalversion'"
+msg "Using CADICAL='$CADICAL_DIR' version '$cadicalversion'"
 
-if [ x"`grep 'bool flip' $CADICAL/src/cadical.hpp`" = x ]
+if [ "$(grep 'bool flip' "$CADICAL_DIR"/src/cadical.hpp)" = "" ]
 then
   [ $flip = no ] || \
     die "no 'CaDiCaL' support for 'bool flip (int lit)' (update version or use '--no-flip')"
@@ -68,14 +66,14 @@ else
   flip=yes
 fi
 
-if [ x"$CXX" = x ]
+if [ "$CXX" = "" ]
 then
   COMPILE="g++ -fPIC -Wall"
 else
   COMPILE="$CXX -fPIC -W"
 fi
 
-if [ x"$AR" = x ]
+if [ "$AR" = "" ]
 then
   LINK="ar"
 else
@@ -94,6 +92,6 @@ fi
 msg "Compiling with '$COMPILE'"
 
 rm -f makefile
-sed -e "s/@COMPILE@/$COMPILE/" makefile.in | sed -e "s/@LINK@/$LINK/" > makefile
+sed -e "s|@COMPILE@|$COMPILE|" makefile.in | sed -e "s|@LINK@|$LINK|" > makefile
 
 msg "Generated 'makefile' (run 'make' to compile)"
diff --git a/generate b/generate
index 4d0d0bb..fc4ab93 100755
--- a/generate
+++ b/generate
@@ -1,16 +1,6 @@
 #!/bin/sh
-die () {
-  echo "generate: error: $*" 1>&2
-  exit 1
-}
-[ -f VERSION ] || die "could not find 'VERSION' file"
-[ -d .git ] || die "could not find '.git' directory"
-[ -f makefile ] || die "could not find 'makefile'"
-VERSION="`cat VERSION`"
-GITID="`git show|head -1|awk '{print $2}'`"
-BUILD="`sed '/^COMPILE/!d;s,^COMPILE=,,;s, -I.*,,' makefile`"
 cat <<EOF
-#define VERSION "$VERSION"
-#define GITID "$GITID"
-#define BUILD "$BUILD"
+#define VERSION "$CADICAL_VERSION"
+#define GITID "$CADICAL_GITID"
+#define BUILD "$(sed '/^COMPILE/!d;s,^COMPILE=,,;s, -I.*,,' makefile)"
 EOF
diff --git a/makefile.in b/makefile.in
index 5152cf9..292ee60 100644
--- a/makefile.in
+++ b/makefile.in
@@ -1,12 +1,12 @@
 COMPILE=@COMPILE@
 LINK=@LINK@
 all: libcadiback.a libcadiback.so
-libcadiback.a: cadiback.o ../cadical/build/libcadical.a makefile
+libcadiback.a: cadiback.o ${CADICAL_DIR}/build/libcadical.a makefile
 	$(LINK) rcs $@ cadiback.o
-libcadiback.so: cadiback.o ../cadical/build/libcadical.so makefile
-	$(COMPILE) -shared -o $@ -L../cadical/build/ -lcadical cadiback.o
-cadiback.o: cadiback.cpp config.hpp ../cadical/src/cadical.hpp makefile
-	$(COMPILE) -c $< -I../cadical/src
+libcadiback.so: cadiback.o ${CADICAL_DIR}/build/libcadical.so makefile
+	$(COMPILE) -shared -o $@ -L${CADICAL_DIR}/build/ -lcadical cadiback.o
+cadiback.o: cadiback.cpp config.hpp ${CADICAL_DIR}/src/cadical.hpp makefile
+	$(COMPILE) -c $< -I${CADICAL_DIR}/src
 config.hpp: generate VERSION makefile
 	./generate > $@
 clean:
