diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 80e4751..d056422 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -20,7 +20,9 @@
 
 include_directories(${PROJECT_SOURCE_DIR})
 include_directories(${CMAKE_CURRENT_BINARY_DIR})
-include_directories(${PROJECT_SOURCE_DIR}/eigen-3.4.0/)
+
+find_package(argparse CONFIG REQUIRED)
+find_package(Eigen3 CONFIG REQUIRED)
 
 if (NOT WIN32)
     add_cxx_flag_if_supported("-Wno-bitfield-constant-conversion")
@@ -55,7 +57,8 @@ set_target_properties(sbva PROPERTIES
 add_executable (sbva-bin main.cpp)
 add_executable (test test.cpp)
 
-target_link_libraries(sbva-bin sbva)
+target_link_libraries(sbva Eigen3::Eigen)
+target_link_libraries(sbva-bin sbva argparse::argparse)
 target_link_libraries(test sbva)
 
 set_target_properties(test PROPERTIES
diff --git a/src/main.cpp b/src/main.cpp
index f1a0fca..961717c 100644
--- a/src/main.cpp
+++ b/src/main.cpp
@@ -28,7 +28,7 @@ THE SOFTWARE.
 #include <vector>
 #include <string>
 #include "sbva.h"
-#include "argparse.hpp"
+#include <argparse/argparse.hpp>
 #include "time_mem.h"
 
 #if defined(__GNUC__) && defined(__linux__)
